working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1836,8 +2049,134 @@ main() {
+checking), a new kind of object that needs a runtime representation and runtime
+type, and new imperative behavior.
+
+It might be too much to fit into a single Dart release. However, it isn't
+necessary to ship every corner of these proposals all at once. If needed for
+scheduling reasons, we could stage it across several releases.
+
+Here is one way it could be broken down into separate pieces:
+
+*   **Records and destructuring.** Record expressions and record types are one
+    of the most-desired aspects of this proposal. Currently, there is no
+    expression syntax for accessing positional fields from a record. That means
+    we need destructuring. So, at a minimum:
+
+    *   Record expressions and types
+    *   Pattern variable declarations
+    *   Record patterns
+    *   Variable patterns
+
+    This would not include any refutable patterns, so doesn't need the changes
+    to allow patterns in switches.
+
+*   **Collection destructuring.** A minor extension of the above is to also
+    allow destructuring the other built-in aggregate types:
+
+    *   List patterns
+    *   Map patterns
+
+*   **Extractors.** I don't want patterns to feel like we're duct taping a
+    functional feature onto an object-oriented language. To integrate it more
+    gracefully means destructuring user-defined types too, so adding:
+
+    *   Extractor patterns
+
+*   **Refutable patterns.** The next big step is patterns that don't just
+    destructure but *match*. The bare minimum refutable patterns and features
+    are:
+
+    *   Patterns in switch statement cases
+    *   Switch case guards
+    *   Exhaustiveness checking
+    *   Literal patterns
+    *   Constant patterns
+    *   Relational patterns (at least `==`)
+
+    The only critical relational pattern is `==` because once we allow patterns
+    in switch cases, we lose the ability to have a bare identifier constant in
+    a switch case.
+
+*   **Type patterns.** The other type-based patterns aren't critical but do
+    make patterns more convenient and useful:
+
+    *   Null-check patterns
+    *   Null-assert patterns
+    *   Cast patterns
+
+*   **Control flow.** Switch statements are heavyweight. If we want to make
+    refutable patterns more useful, we eventually want:
+
+    *   Switch expressions
+    *   Pattern-if statements
+
+*   **Logic patterns.** If we're going to add `==` patterns, we may as well
+    support other Boolean infix operators. And if we're going to support the
+    comparison operators, then `&` is useful for numeric ranges. It's weird to
+    have `&` without `|` so we may as well do that too (and it's useful for
+    switch expressions). Once we have infix patterns precedence comes into play,
+    so we need parentheses to control it:
+
+    *   Relational patterns (other than `==`)
+    *   Or patterns
+    *   And patterns
+    *   Grouping patterns
+
+*   **Type patterns.** This is a mostly orthogonal feature, though a useful one
```

Comment:
Also the title in line 2121, covering something else.

Reply:
Done.